### ISSUE TITLE
bisync: beta time improvements

### ISSUE THREAD

**Issue by ivandeex** [rclone/rclone#5674]
### Intro

The bisync pull request #5164 appeared too large to fulfil in a single rclone release.

Plan is to merge it right after the rclone v1.57 release. The command will be available in the beta builds for a limited circle of beta testers until official release with rclone v1.58.
During the beta period we can merge a number of postponed items, **in particular breaking changes**.

This ticket will keep track of the most beta time changes. Large and critical changes will be split into separate tickets.
High priority items reflect changes breaking user experience and must be completed before 1.58. Medium priority items are required but not guaranteed by 1.58. Optional and low priority items will be implemented on best effort basis. Optional items don't even have dedicated tickets yet.

### High priority

- [ ] Track critical flags in saved sessions - #5678
- [ ] Support imprecise or absent modtimes - #5679
- [ ] Use hashsums, honor `--checksum` - #5683

### Medium priority

- [ ] Check file sizes - #5684
- [ ] Respect `--backup-dir` - #5685
- [ ] Align `--max-delete` semantics - #5680
- [ ] Optimize resync - #5681
- [ ] Track renames - #5690

### Low priority

- [ ] Handle deletion conflicts - #5691
- [ ] Honor `--interactive` and `--auto-confirm`
- [ ] Recover after transient failures - #5676
- [ ] Use **trie** for listings to reduce footprint - #5686
- [ ] Maintain session history - #5692
- [ ] Handle Google Docs better - #5696

### Optional items

- [ ] Parallelize move/rename operations
- [ ] Honor special `.rcignore` files
- [ ] Revisit progress display for bisync
- [ ] Handle case-insensitive filenames. Respect `--ignore-case` and `--ignore-case-sync` ([rcsync#54](https://github.com/cjnaz/rclonesync-V2/issues/54), [#4854](https://github.com/rclone/rclone/issues/4854))
- [ ] Synchronize empty dirs if supported by backend ([rcsync#63](https://github.com/cjnaz/rclonesync-V2/issues/63))
- [ ] Fix tests for Google Photos ([rcsync#53](https://github.com/cjnaz/rclonesync-V2/issues/53), [G.Photos issues](https://github.com/rclone/rclone/issues?q=is%3Aopen+is%3Aissue+label%3A%22Remote%3A+Google+Photos%22))
- [ ] Enable CI tests with OwnCloud (according to @cjnaz it's stable)
- [ ] Test should handle Dropbox "touch quirks" in a generic way

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ivandeex** [comment]
## rename RCLONE_TEST

...

**Comment by ivandeex** [comment]
## trie

...

**Comment by ivandeex** [comment]
## investigate bisync chains

...

**Comment by ivandeex** [comment]
## --track-renames

...

**Comment by ivandeex** [comment]
## --backup-dir

https://github.com/rclone/rclone/pull/5164#pullrequestreview-729822158

Will users be able to use something like `--backup-dir` or `--suffix`? I can see users wanting that to prevent data loss. Maybe `--suffix` works already and you could filter the backups out using the filtering? Perhaps this should be automatic.

**Comment by ivandeex** [comment]
## Revisit --resync

Optimize --resync as copy+copy ([rcsync#66](https://github.com/cjnaz/rclonesync-V2/issues/66))

**Comment by ivandeex** [comment]
## Guard against dangerous filters#5673 

### https://github.com/rclone/rclone/pull/5164#issuecomment-842407758 (JWink3101)

Why are the filters different? The only one of concern is `--exclude-if-present` which can cause issues with bi-directional sync. The following is the note I include in my readme about them:

> Filters work well since they are applied to both sides. This means that changing a filter will make the files look like it was deleted on *both* sides and nothing will happen.
> 
> However, the [`--exclude-if-present`][eifp] filter is **very dangerous** if the excluded file is added on one side only *after* files in the directory are in sync. It will cause rclone to skip the directory on one side and appear like it was deleted. 
> 
> There *are* safe ways to use `--exclude-if-present`. For example, you can place exclude file in place on *both* sides before syncing. (Or placing it without adding the filter, syncing, and then adding the filter). Or, if the files have *never* been synced before (i.e. it's a new directory with the exclude file) then nothing bad will happen.

This isn't a new problem in bisync from regular rclone though I do see it amplified for it. Again this goes back to the philosophical idea of a sync-config vs CLI flags. The reason I chose not to use the CLI flags was because I wanted it very clear that a "sync job" is not a one-off.

**Comment by cjnaz** [comment]
Related to "Honor special .rcignore files"

[https://github.com/cjnaz/rclonesync-V2/issues/76](https://github.com/cjnaz/rclonesync-V2/issues/76)

**Comment by cjnaz** [comment]
```
Enable CI tests with OwnCloud (according to @cjnaz it's stable)
```

What's the issue???  Say more.

**Comment by nielash** [comment]
This list is very old -- see the [bisync project](https://github.com/rclone/rclone/projects/6), [label](https://github.com/rclone/rclone/issues?q=is%3Aopen+sort%3Aupdated-desc+label%3Abisync+), and [milestones](https://github.com/rclone/rclone/milestones) for a more up-to-date view of the roadmap. (But I will keep this issue open at least until I verify that everything on this list is accounted for elsewhere.)
